### ISSUE TITLE
Unexpected behaviour when setting properties in module vars at module level

### ISSUE THREAD

**Issue by spalladino** [crystal-lang/crystal#4051]
When assigning a property of a module variable at module level, that assignment only affects the module variable of the module itself, and not in any class that includes the module. See below:

```crystal
class Test
  property test = 0
end

module Foo
  @@var = Test.new
  @@var.test = 5
  
  def var
    @@var
  end
  
  def self.var
    @@var
  end
end

class Bar
  include Foo
end

bar = Bar.new
p bar.var.test => 0
p Foo.var.test => 5
```

On Crystal 0.20.5, OSX El Capitan.

**Comment by spalladino** [comment]
This happens because code written at module level is executed only in the context of the module, and **not** in the context of every class where it is included. Since every class gets its copy of a `@@var`, only the one in the module gets modified.

Removing the bug tag until we agree if this behaviour is expected, or we want to review what can be written at module definition level.

**Comment by asterite** [comment]
As a workaround you can do:

```crystal
module Foo
  @@var = begin
    var = Test.new
    var.test = 5
    var
  end
end
```

Or:

```crystal
module Foo
  @@var = Test.new.tap { |var| var.test = 5 }
end
```

**Comment by r-obert** [comment]
Looks like a bug to me. 
I don't think it should matter where the code executes, because conceptually `@@var` is a reference to the same object. Its properties should be the same everywhere.

**Comment by asterite** [comment]
@r-obert No, class vars are inherited and each type has its own copy

**Comment by drosehn** [comment]
I think it's significant to note that the example calls `Foo.var.test` not `Bar.var.test`.  It has to, because there is no `Bar.var.test`.  The `self` in `def self.var` is always `Foo`, and not the class-name of whatever `Foo` is included in.

**Comment by drosehn** [comment]
Here is a program which maybe makes it more obvious, and also let's you experiment with the work-arounds that asterite wrote:
```cr
class Test
  property test = 0
end

module Foo
  @@var = Test.new
  @@var.test = 5

  @@vas : Test = Test.new ;   @@vas.test = 7
# @@vas : Test = Test.new.tap { |var| var.test = 15 }
# @@vas : Test = begin ; var = Test.new ; var.test = 25 ; var ; end
 
  def var
    printf "instc: %s var=%s vas=%s\n", self.class, @@var.inspect, @@vas.inspect
    @@var
  end
  
  def self.var
    printf "class: %s var=%s vas=%s\n", self.name, @@var.inspect, @@vas.inspect
    @@var
  end

  def vas
    @@vas
  end
  def self.vas
    @@vas
  end
end

class Bar
  include Foo

  def self.c_vas
    printf "c_vas: %s var=%s vas=%s\n", self.name, @@var.inspect, @@vas.inspect
    @@vas
  end
end

bar = Bar.new
printf "= Foo.var=%-2d bar.var=%d\n  Foo.vas=%-2d bar.vas=%-2d Bar.c_vas=%d\n",
       Foo.var.test, bar.var.test,
       Foo.vas.test, bar.vas.test, Bar.c_vas.test
```
 produces:
```cr
class: Foo var=#<Test:0x100ff2fd0 @test=5> vas=#<Test:0x100ff2fc0 @test=7>
instc: Bar var=#<Test:0x100ff2fa0 @test=0> vas=#<Test:0x100ff2f90 @test=0>
c_vas: Bar var=#<Test:0x100ff2fa0 @test=0> vas=#<Test:0x100ff2f90 @test=0>
= Foo.var=5  bar.var=0
  Foo.vas=7  bar.vas=0  Bar.c_vas=0
```

And then comment-and-uncomment lines to change which tactic you want to try for initializing `@@vas`:
```cr
class: Foo var=#<Test:0x1045d7fd0 @test=5> vas=#<Test:0x1045d7fc0 @test=15>
instc: Bar var=#<Test:0x1045d7fa0 @test=0> vas=#<Test:0x1045d7f90 @test=15>
c_vas: Bar var=#<Test:0x1045d7fa0 @test=0> vas=#<Test:0x1045d7f90 @test=15>
= Foo.var=5  bar.var=0
  Foo.vas=15 bar.vas=15 Bar.c_vas=15
```
or:
```cr
class: Foo var=#<Test:0x104a3efd0 @test=5> vas=#<Test:0x104a3efc0 @test=25>
instc: Bar var=#<Test:0x104a3efa0 @test=0> vas=#<Test:0x104a3ef90 @test=25>
c_vas: Bar var=#<Test:0x104a3efa0 @test=0> vas=#<Test:0x104a3ef90 @test=25>
= Foo.var=5  bar.var=0
  Foo.vas=25 bar.vas=25 Bar.c_vas=25
```

**Comment by drosehn** [comment]
I'm too tired to really think straight right now, but I suspect another work-around would be for instance-methods to call method-level class-methods for the value of an `@@variable`.  Eg:
```cr
module Foo
  @@mod_var = Test.new
  @@mod_var.test = 5
  
  def var
    Foo.mod_var
  end
  
  def self.var
    @@mod_var
  end
end
```
It wouldn't surprise me if that doesn't really make any sense to do, but it seemed like an interesting tactic to me.  I guess it depends whether you want the `@@variable` in a module to be isolated to that module, or that it can be referenced directly by any methods in a class which includes the module.

**Comment by r-obert** [comment]
@asterite But `@@var.test` is initialised to 5 before Bar is defined, so the copy should be the same as the object initialised in Foo.

**Comment by drosehn** [comment]
It might be misleading to think of it as "a copy".  The instance of `@@test` in `module Foo` is a completely different variable (which exists at a different location in memory) than the instance of `@@test` in `class Bar`.  If I understand this right, it is not created by copying the first instance.

I must admit that I keep going back-and-forth in whether I think this behavior makes sense.  I do see how it gives the programmer two different ways to handle `@@-variables`, and that both of those could be useful, but I'm not sure it "makes sense".  I think I need to write an even more elaborate example to be sure that I understand what is going on here.  😄

**Comment by drosehn** [comment]
Consider this example (using `Crystal 0.21.0`):
```cr
class Test
  @@test_counter = 0
  getter test : Int32

  def initialize
    @test = @@test_counter
    @@test_counter += 1
    printf "new Test object: #%d\n", @test
  end

  def test=(newval : Int32)
    printf "    Changing @test from %d to %d\n", @test, newval
    @test = newval
  end
end

module Foo
  @@var = Test.new
  @@var.test = 5

  def var
    printf "    instc: %s var=%s\n", self.class, @@var.inspect
    @@var
  end
  
  def self.var
    printf "    class: %s var=%s\n", self.name, @@var.inspect
    @@var
  end
end

class Bar
  include Foo
end
class Baz
  include Foo
  @@var.test = 7
end

bar = Bar.new
baz = Baz.new
printf "= Foo.var=%-2d bar.var=%d baz.var=%d\n", Foo.var.test,
    bar.var.test, baz.var.test
```
which produces:
```cr
new Test object: #0
    Changing @test from 0 to 5
new Test object: #1
    Changing @test from 1 to 7
    class: Foo var=#<Test:0x1094cdfd0 @test=5>
new Test object: #2
    instc: Bar var=#<Test:0x1094cdf90 @test=2>
    instc: Baz var=#<Test:0x1094cdfc0 @test=7>
= Foo.var=5  bar.var=2 baz.var=7
```
This shows that three separate `@@var`s are created; one each for `module Foo`, `class Bar`, and `class Baz`.  So if you're looking at `module Foo`, it seems like the `@@var = Test.new` line is executed three times, but the `@@var.test = 5` is executed only once.

I guess I don't understand why `bar.var=2` instead of `1`, but I also feel that detail is not important.  This would be pretty strange code to write for a real-life application.

I have no strong feeling on whether the rest of the behavior makes sense, but I hope this example helps to show the interactions between classes and included-modules.  Or maybe all of this behavior was already obvious to everyone but me!

**Comment by asterite** [comment]
OK, here's another example:

```crystal
module Foo
  @@var = 1

  puts "Var is: #{@@var}"
end

class Bar
  include Foo
end
```

Should the `puts` be executed twice here, once for Foo and one for Bar? I guess the answer is pretty obvious: no. So the answer to this issue is:

1. Something like `@@var = exp` is a class var initializer: it's executed once per type in the hierarchy that defines the var.
2. Anything else is a top-level expression that's executed as soon as the program reaches that point.

In the original snippet we have:

- `@@var = Test.new` -> this is a class var initializer, executed once per type
- `@@var.test = 5` -> this is a top-level expression, executed once, and only for Foo (because it's in the scope of Foo)

We could maybe disallow top-level code inside classes to fix this issue. Then you'll get an error on the `@@var.test = 5` line, and you'll be forced to write it with `begin ... end`. This is only an issue because you `@@var = exp`, which looks like a top-level expression, is a class var initializer. It's similar to how `@var = exp` is an instance var initializer, not a top-level expression.

**Comment by drosehn** [comment]
The following is not a request for changing anything, but is only a question out of curiosity:
> Should the `puts` be executed twice here, once for Foo and one for Bar? I guess the answer is pretty obvious: no.

What is the downside with it being executed in both places?  I'd say the odd thing about that example is that the programmer wrote a `puts` at the top-level of the module.  But if they did that, then why not execute it?

To ask this a different way:  Let's say that there were two languages already written, and one of them did execute commands which are at the top-level of modules, and the other did not.  So, all the hard work of writing the language is already done.  If I then come along and want to decide which of those languages to pick for my own project, then why would I pick one language over the other?

In my own mind, either way is fine as long as programmers realize what's happening.  But I don't understand the statement "the answer is pretty obvious: no".  I do not understand what is obvious about that answer.

**Comment by drosehn** [comment]
Here is another question.  In this case my thinking goes in the exact opposite direction of my previous question.  So, yeah, I often get into arguments with myself.  In any case:

For me, the most unexpected thing shown by this discussion was that a `@@var = Someclass.new` line would be executed when processing the top-level of a `module`.  The `initialize` method for a `Someclass` object might perform all kinds of work while creating an object for the initial `module` definition, only to have that object just sit around unused if the module itself is only used by being included in other `class` definitions.  All the instance-methods for a class which included the module will reference the object created in the scope of that class, not the object created in the scope of the top-level module.

I suspect there are some situations where I depend on that behavior without realizing it, but I was still surprised to see what that behavior means in situations like the ones shown in this discussion.

**Comment by r-obert** [comment]
Yeah, if an error is raised that suggests begin/end or tap, then at least a subtle bug could be avoided. 

As to your original example, I would expect it to work like Ruby :-) So, include would not act as some sort of eval that would call "puts" a second time. When making copies, I would look at the table of class variables of the superclass, and make a copy through clone/dup/something else. But how it works now seems to be the class var initialisation is re-evaluated in the scope of each type.
